chore: new contributing guidelines (#3485)
adobe/spectrum-web-components: CONTRIBUTING.md

-   Check the [Issue Discussions](https://github.com/adobe/spectrum-web-components/issues).
-   You might be able to find the cause of the problem and fix things yourself. Most importantly, check if you can reproduce the problem [in the latest version of Spectrum Web Components](https://studio.webcomponents.dev/workspace/adobe).

-   Perform a [search](https://github.com/adobe/spectrum-web-components/issues) to see if the problem has already been reported. If it has and the issue is still open, add a comment to the existing issue instead of opening a new one.

### Report bugs using Github's [issues](https://github.com/adobe/spectrum-web-components/issues)

We use GitHub issues to track public bugs. Report a bug by [opening a new issue](https://github.com/adobe/spectrum-web-components/issues/new/choose); it's that easy!

### Write bug reports with detail, background, and sample code

Create an issue on the repository and provide the following information by filling in [the template](https://github.com/adobe/spectrum-web-components/blob/main/.github/ISSUE_TEMPLATE/bug_report.yaml).

-   A quick summary and/or background
-   Steps to reproduce -
    -   Please Be specific!

>     -   Optionally, give sample code [here](https://studio.webcomponents.dev/workspace/adobe)

Reviewer: ```suggestion
    -   Reproduce your issue [here](https://studio.webcomponents.dev/workspace/adobe)
```

We don't want this part to be optional!!

> -   What you expected would happen

Reviewer: ```suggestion
-   What you would expect to happen
```

-   What actually happens

> -   Notes (possibly including why you think this might be happening, or stuff you tried that didn't work)

Reviewer: ```suggestion
-   Notes (for example, why you think this might be happening, or stuff you tried that didn't work)
```

## Do you intend to add a new feature or change an existing one?

-   Suggest your change in the swc-core mailing list and start writing code.

-   Do not open an issue on GitHub until you have collected positive feedback about the change. You can start on [Adding your idea here](https://github.com/adobe/spectrum-web-components/discussions/categories/ideas) under Ideas

> -   Once your Idea or request is submitted, you can start creating an [issue](https://github.com/adobe/spectrum-web-components/issues/new/choose) with the New Component or a Feature Request template.

Reviewer: ```suggestion
-   Once your Idea or request is submitted, you can start creating an [issue](https://github.com/adobe/spectrum-web-components/issues/new/choose) with the New Component or a Feature Request template.
```
```suggestion
-   Once your Idea or request is submitted, you can create an [issue](https://github.com/adobe/spectrum-web-components/issues/new/choose) with the New Component or a Feature Request template.
```

-   Please make sure that your PR follows our [best practices](https://opensource.adobe.com/spectrum-web-components/migrations/2021-8-11/#using-lit%402.0-inside-of-lit-html-and%2For-litelement) and [styleguide](https://spectrum.corp.adobe.com/page/design-tokens/#Usage-guidelines) before you submit for review

## Branching Strategy

### Branch Naming Convention:

-   Use `[username]/[issue]` format for branches.
-   `[username]` identifies the contributor.

> -   `[issue]` refers to the associated issue or feature number.

Reviewer: Thank you for adding this!!!!

For e.g: Be descriptive after the /, like `john-doe/123-fix-bug`.

> Please follow these steps to have your contribution considered by the maintainers

Reviewer: I don't know if this line is necessary?

# How to Contribute to Spectrum Web Components

First off, thanks for taking the time to contribute!

The following is a set of guidelines for contributing to Spectrum Web Components which is hosted in the [Adobe Spectrum Web Components](https://github.com/adobe/spectrum-web-components) on GitHub. These are mostly guidelines, not rules. Use your best judgment, and feel free to propose changes to this document in a pull request.

> ## I don't want to read this whole thing I just have a question!!!

Reviewer 2: Maybe rename this to just TL;DR; or something about the content more directly. This title and the next effectively says "have a question".

Reviewer 2: Or something like "Things to know"?

Reviewer: I like TL;DR!

The following is a set of guidelines for contributing to Spectrum Web Components which is hosted in the [Adobe Spectrum Web Components](https://github.com/adobe/spectrum-web-components) on GitHub. These are mostly guidelines, not rules. Use your best judgment, and feel free to propose changes to this document in a pull request.

## Code Of Conduct
## I don't want to read this whole thing I just have a question!!!

> This project adheres to the Adobe [code of conduct](CODE_OF_CONDUCT.md). By participating,

Reviewer: I think we should still keep this code of conduct stuff in the documentation.

Author: Sure. Added back

> -   [Github Discussions, the official Spectrum Web Components message board](https://github.com/adobe/spectrum-web-components/discussions)
> -   [Spectrum Web Components FAQ](https://github.com/adobe/spectrum-web-components/discussions/categories/q-a)

Reviewer 2: I like both these but maybe it would help to briefly describe more? 
```suggestion
-   [Spectrum Web Components FAQ](https://github.com/adobe/spectrum-web-components/discussions/categories/q-a): Your question may have been asked before, take a quick look at the FAQ to see previous questions that may help you.
-   [Github Discussions, the official Spectrum Web Components message board](https://github.com/adobe/spectrum-web-components/discussions): If you don't see a previously asked question, feel free to reach out to the team in the message board.
```

> **Note:** [Please don't file an issue to ask a question.](https://github.com/adobe/spectrum-web-components/discussions) You'll get faster results by using the resources below.

## Have A Question?
We have an official message board with a detailed FAQ and where the community chimes in with helpful advice if you have questions.

Start by filing an issue. The existing committers on this project work to reach
consensus around project direction and issue solutions within issue threads
(when appropriate).
-   [Github Discussions, the official Spectrum Web Components message board](https://github.com/adobe/spectrum-web-components/discussions)
-   [Spectrum Web Components FAQ](https://github.com/adobe/spectrum-web-components/discussions/categories/q-a)

## Contributor License Agreement
## How Can I Contribute?

### Contributor License Agreement

All third-party contributions to this project must be accompanied by a signed contributor
license agreement. This gives Adobe permission to redistribute your contributions
as part of the project. [Sign our CLA](http://opensource.adobe.com/cla.html). You
only need to submit an Adobe CLA one time, so if you have submitted one previously,

> you are good to go!

Reviewer 2: This seems like important context about why you need to sign the CLA? Maybe we should keep some of this?

Reviewer: I agree. The removed code of conduct and the emails for violations could probably be added here, as well.

Author: Done

> Do not open up a GitHub issue if the bug is a security vulnerability in Rails, and instead to refer to our [security policy](https://helpx.adobe.com/security/alertus.html).

Reviewer: Is this an internal link? If so, we may not be able to use it here.

Reviewer: I saw that it was in the documentation previously... so it might be fine?

> -   **Check the [Issue Discussions](https://github.com/adobe/spectrum-web-components/issues).** You might be able to find the cause of the problem and fix things yourself. Most importantly, check if you can reproduce the problem [in the latest version of Spectrum Web Components](https://studio.webcomponents.dev/workspace/adobe)

Reviewer 2: ```suggestion
-   **Check the [Issue Discussions](https://github.com/adobe/spectrum-web-components/issues).** You might be able to find the cause of the problem and fix things yourself. Most importantly, check if you can reproduce the problem [in the latest version of Spectrum Web Components](https://studio.webcomponents.dev/workspace/adobe). This will isolate the issue outside your particular use case and help us discuss the issue more directly.
```

> -   **Check the [Issue Discussions](https://github.com/adobe/spectrum-web-components/issues).** You might be able to find the cause of the problem and fix things yourself. Most importantly, check if you can reproduce the problem [in the latest version of Spectrum Web Components](https://studio.webcomponents.dev/workspace/adobe)

Reviewer: Please keep the commitlint stuff here as well. It's a useful resource for new contributors!

Bugs are tracked as [GitHub issues](https://guides.github.com/features/issues/). Create an issue on the repository and provide the following information by filling in [the template](https://github.com/adobe/spectrum-web-components/blob/main/.github/ISSUE_TEMPLATE/bug_report.yaml).

[IMPORTANT] Explain the problem and include additional details to help maintainers reproduce the problem:

-   **Describe the exact steps which reproduce the problem** in as many details as possible.
-   **Provide specific examples to demonstrate the steps**. If you're providing snippets in the issue, use [Playground](https://studio.webcomponents.dev/workspace/adobe).
-   **Explain which behavior you expected to see instead and why.**

> -   **Include screenshots and animated GIFs** which show you following the described steps and clearly demonstrate the problem.

Reviewer: ```suggestion
-   **Optionally, Include screenshots and animated GIFs** which show you following the described steps and clearly demonstrate the problem.
```

I think the gifs/screenshots are important, but not AS important as the reproduction in webcomponents.dev, so I would say it's optional 😅

## Do you intend to add a new feature or change an existing one?

> -   Suggest your change in the swc-core mailing list and start writing code.

Reviewer: Can you elaborate what the swc-core mailing list is?

Reviewer: Bumping this question again

> -   Do not open an issue on GitHub until you have collected positive feedback about the change. You can start on [Adding your idea here](https://github.com/adobe/spectrum-web-components/discussions/categories/ideas) under Ideas

Reviewer: I think it's fine if people want to request features through our issue template, but it probably makes our issues more organised if we keep those to the discussions. 🤔

> -   Once your Idea or request is submitted you can start creating a [issue](https://github.com/adobe/spectrum-web-components/issues/new/choose) with New Component or a Feature Request

Reviewer: ```suggestion
-   Once your Idea or request is submitted, you can start creating an [issue](https://github.com/adobe/spectrum-web-components/issues/new/choose) with the New Component or a Feature Request template.
```

> -   Please make sure that your PR follows our best practices and styleguide before you submit for review

Reviewer 2: Are these spaces between bullet points intentional?

Author: Nope just had extra tabs.

> -   Please make sure that your PR follows our best practices and styleguide before you submit for review

Reviewer: ```suggestion
-   Please make sure that your PR follows our best practices and styleguide before you submit for review
```

Some links to other documentation about best practises, the styling guide, commitlint standards, etc would be great here.

> ## Create your first Pull Request

Reviewer 2: Should we add a recommendation about how to branch/fork/whatever? What is the best practice for creating a branch both in naming and location?

Reviewer 2: … for example, should you fork the repo? create a branch in the repo? etc. Normally a fork would be recommended but does that work with testing setup?

Reviewer 3: If you can make a branch, that is preferable. A branch has full access to the CI, especially VRTs; maybe doing some work to see if that could be made available for forks could be useful here. However, if you use a fork, and aren't making visible changes, you get the unit test suite in that context.

Reviewer: An important note: We also use the branch syntax of [username]/[issue]

The part after the / is kind of... loose. I think having your username is more important so we know who did the work.

Please follow these steps to have your contribution considered by the maintainers:

1. Follow all instructions in [the template](PULL_REQUEST_TEMPLATE.md)
2. After you submit your pull request, verify that all [status checks](https://help.github.com/articles/about-status-checks/) are passing.

## Git Commit Messages

We maintain a specific standard of commit messages while committing to the respository. See [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)

## Do you want to contribute to the Spectrum Web Components documentation?

-   Please read [Contributing Guidelines for documentation](https://github.com/adobe/spectrum-web-components#documentation)

## Pull Request Labels

> Specific to your contribution and the status we would encourage you to use PR labels such as `WIP`, `triage` , `under-review`, `testing`

Reviewer: I think other. than "WIP", labels aren't necessary from my perspective. As long as the PR is linked to an issue, that helps, and you can flag a PR as a draft. I think it's good to specify reviewers and notify them when your PR is ready for review. Let me know what you think.

Author: Yea make sense. A little bit of overhead i think. Lets not add this specification. I think it would be self explanatory.

Thanks! :heart: :heart: :heart:

> Spectrum Team

Reviewer: ```suggestion
The Spectrum Web Components Team
```
